### PR TITLE
HPCC-19306 Support OpenSSL 1.1 and Ubuntu 18.04

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Compiled

### CODE REVIEW

**Comment by richardkchapman** [comment]
@RussWhitehead Please review

File: system/security/securesocket/securesocket.cpp [hunk 1]
```diff
@@ -40,6 +40,7 @@
 #endif
 
 //openssl
+#define OPENSSL_USE_DEPRECATED 1
```
Comment (richardkchapman) [inline system/security/securesocket/securesocket.cpp]: Actually this didn't seem to help so I should probably remove it

File: system/security/securesocket/securesocket.cpp [hunk 2]
```diff
@@ -1320,8 +1336,11 @@ class CRsaCertificate : implements ICertificate, public CInterface
 
         X509_free(x509);
         EVP_PKEY_free(pkey);
+        RSA_free(rsa);
```
Comment (richardkchapman) [inline system/security/securesocket/securesocket.cpp]: I think we were probably leaking this before?

File: system/security/securesocket/securesocket.cpp [hunk 3]
```diff
@@ -1228,7 +1235,16 @@ class CRsaCertificate : implements ICertificate, public CInterface
         if ((x509=X509_new()) == NULL)
             throw MakeStringException(-1, "can't create X509 structure");
 
-        RSA *rsa=RSA_generate_key(m_bits, RSA_F4, NULL, NULL);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+        RSA *rsa = RSA_generate_key(m_bits, RSA_F4, NULL, NULL);
```
Comment (RussWhitehead) [inline system/security/securesocket/securesocket.cpp]: Should throw if NULL returned here and RSA_new() on 1241

**Review by RussWhitehead — CHANGES_REQUESTED** [review]

**Review by RussWhitehead — APPROVED** [review]
Looks good to go

**Comment by richardkchapman** [comment]
@ghalliday Please merge

File: system/security/securesocket/securesocket.cpp [hunk 4]
```diff
@@ -1677,7 +1723,9 @@ SECURESOCKET_API int signCertificate(const char* csr, const char* ca_certificate
     PEM_write_bio_X509(cmem, cert);
     readBio(cmem, certificate);
 
-    CRYPTO_mem_leaks(bio_err);
+#ifndef OPENSSL_NO_CRYPTO_MDEBUG
+        CRYPTO_mem_leaks(bio_err);
```
Comment (ghalliday) [inline system/security/securesocket/securesocket.cpp]: whitespace: indentation off

**Comment by ghalliday** [comment]
@richardkchapman please squash first.  One minor whitespace comment.

**Comment by richardkchapman** [comment]
@ghalliday Back to you

**Pull request merged**